gto-support-db-coroutines/src/main/kotlin/org/ccci/gto/android/common/db/CoroutinesFlowDao.kt

```diff
@@ -1,17 +1,36 @@
 package org.ccci.gto.android.common.db
 
 import androidx.annotation.AnyThread
-import androidx.lifecycle.asFlow
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+
+interface CoroutinesFlowDao : CoroutinesDao, Dao {
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private fun invalidationFlow(types: Collection<Class<*>>) = when {
+        types.isEmpty() -> flowOf(Unit)
+        else -> callbackFlow {
+            val callback = Dao.InvalidationCallback { if (it in types) trySendBlocking(it) }
```

Comment:
So does ```trySendBlocking``` prevent the producer from sending any more data to the consumer until the consumer either returns that it was successful or failed in receiving the data?

Reply:
this bridges from synchronous to asynchronous within the flow and triggers `send`.

Basically the `Dao.InvalidationCallback` is called from a synchronous context (the thread that just updated the database), we need to provide the class that was just invalidated to our flow, but the flow could be still processing the previous value.

From a blocking context we have the following 3 options:
1. `trySend(it)` - This will `send` the value immediately if the flow is in a state where it can accept a value. If the flow is not ready to accept a value, this will return false and not actually send the value.
2. `runBlocking { send(it) }` - this will block the current thread until the asynchronous `send` is processed.
3. `trySendBlocking(it)` - this does the same as 2, but in a way that covers more edge cases related to exceptions and also avoids the overhead of `runBlocking` if possible.

From a correctness standpoint 1 could drop invalidations that should actually be processed, so we can't go with that. 2 and 3 would both be correct in this scenario, but 3 is a more robust version of 2.

Reply:
This comment also made me realize I accidentally dropped a `.conflate()` call on the invalidationFlow. `conflate()` will discard previous unprocessed items in the flow when a new item shows up. In this case `conflate()` is a good thing.

In the scenario that sync logic runs and updates 10 tools in the database, that would cause the InvalidationCallback to be triggered 10 times with `Tool::class.java`. We should only need to evaluate a single query after the last invalidation fires to get updated data, not the same query 10 times.